### PR TITLE
TabControl rework

### LINKED ISSUE #536: Binding to TabControl.Items doesn't work.
`TabControl` is a special snowflake (i.e. I hate it) and the way that it works makes it really difficult to allow `DataTemplates` that allow automatic creation of tabs from `Items`.

Need to work out how we can fix this, or just :fire: it and tell people to use a `TabStrip` and a `ContentControl`/`Carousel` (not really an option unfortunately).

### PR DESCRIPTION
**Pull request by Gillibald**
- What does the pull request do?
It introduces a reworked TabControl that supports data bound creation of tabs
- What is the current behavior?
The current implementation just supports static data
- What is the updated/expected behavior with this PR?
With the new TabControl you can use a HeaderTemplate and a ContentTemplate for a TabItem to change the way they look. This also works for data bound objects. 
- How was the solution implemented (if it's not obvious)?
The control now uses an ItemsPresenter and a ContentPresenter to arrange objects. It no longer uses TabStrip or Carousel. TabItem got a default template to define the look.

Note: breaking change for styles that were used to customize tabs (TabStripItem)

Fixes #536

### CODE REVIEW

**Comment by grokys** [comment]
Not sure if this is still WIP, but just taking a look through it and here are some random thoughts/questions from taking a quick look through the code:

- It's definitely an improvement on what we have right now!
- Is there no way to use `TabStrip` for the tab strip on the control? Currently the tabs are styled completely differently to the `TabStrip` items. Would be nice if styling one would style both.
- Are transitions going to be supported on the "content" area? I assume not because the "content" part is just a single `ContentPresenter`. Tbh that's fine with me if so, if one requires transitions we can say to use a `TabStrip` and a `Carousel`.
- I think the `TabControl` template doesn't need to be repeated for the different `TabStripPlacements` - a selector could be used to set the `DockPanel.Dock` property

**Comment by Gillibald** [comment]
> I think the TabControl template doesn't need to be repeated for the different TabStripPlacements - a selector could be used to set the DockPanel.Dock property

Thats right but there was a bug that prevented me from doing that. My initial implementation just had a selector for each dock direction and only updated the Dock property. That solution didn't work because on TabStripPlacement change the style didn't get applied.

**Comment by Gillibald** [comment]
I can implement a TransitionalContentPresenter that supports a transition for the content on change if thats a must have.

**Comment by grokys** [comment]
Apologies for the delay on this @Gillibald - I was going to wait until 0.7 was out, but that seems to be taking more time than expected. I will try to do a proper review in the coming days, and we'll maybe get it in for 0.7.

**Comment by Gillibald** [comment]
No hurry. Let's make sure this is a solid solution for the usage of data bound items.

**Comment by igorstefus** [comment]
Sorry took me some time to test it. Templates are loading and control is visible in the header so it seems to work just fine. When will this control become part of the main build? It is definitely better then one we have. 

Any way thanks for all the hard work.

**Comment by igorstefus** [comment]
Sorry took me some time to test it. Control is loading template just fine and it is visible in the header. This tab control is much better then one we have now. When will this be part of the main build? :).

Any way thanks for all hard work.

**Comment by Gillibald** [comment]
@igorstefus Thanks for testing. That's good news. We just want to make sure the new TabControl works as expected. Still needs proper review.

File: src/Avalonia.Controls/ContentControl.cs [hunk 1]
```diff
@@ -67,7 +67,7 @@ public IDataTemplate ContentTemplate
         {
             get { return GetValue(ContentTemplateProperty); }
             set { SetValue(ContentTemplateProperty, value); }
-        }
+        }
```
Comment (grokys) [inline src/Avalonia.Controls/ContentControl.cs]: Nit: whitespace change.

File: src/Avalonia.Controls/Generators/TabItemContainerGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            if (tabItem.HeaderTemplate == null)
+            {
+                tabItem[~HeaderedContentControl.HeaderTemplateProperty] = Owner[~ItemsControl.ItemTemplateProperty];
+            }
+
+            if (tabItem.Header == null)
+            {
+                if (item is IHeadered headered)
+                {
+                    if (tabItem.Header != headered.Header)
```
Comment (grokys) [inline src/Avalonia.Controls/Generators/TabItemContainerGenerator.cs]: Is this check needed? We've established that `tabItem.Header` is null on line 29.

File: src/Avalonia.Controls/TabControl.cs [hunk 3]
```diff
@@ -83,67 +100,82 @@ public Dock TabStripPlacement
...
+            get { return GetValue(ContentTemplateProperty); }
+            set { SetValue(ContentTemplateProperty, value); }
         }
 
-        protected override void OnTemplateApplied(TemplateAppliedEventArgs e)
+        /// <summary>
+        /// Gets or sets the currently selected content.
+        /// </summary>
+        /// <value>
+        /// The content of the selected.
```
Comment (grokys) [inline src/Avalonia.Controls/TabControl.cs]: This doesn't make much sense: the content of the selected what? I'm guessing it should be "The content of the selected tab"?

File: src/Avalonia.Controls/TabControl.cs [hunk 4]
```diff
@@ -83,67 +100,82 @@ public Dock TabStripPlacement
...
-            var content = o as IContentControl;
+            base.OnTemplateApplied(e);
+
+            ItemsPresenterPart = e.NameScope.Find<ItemsPresenter>("PART_ItemsPresenter");
 
-            if (content != null)
+            if (ItemsPresenterPart == null)
             {
-                return content.Content;
+                throw new NotSupportedException("ItemsPresenter not found.");
```
Comment (grokys) [inline src/Avalonia.Controls/TabControl.cs]: You can use the [`INameScope.Get`](https://github.com/AvaloniaUI/Avalonia/blob/dceabb95d99baacc431df7a0a195c905699fc1c1/src/Avalonia.Styling/Controls/NameScopeExtensions.cs#L48) extension method to throw an exception if the requested part isn't found.

File: src/Avalonia.Themes.Default/TabControl.xaml [hunk 5]
```diff
@@ -1,50 +1,64 @@
...
-        </ItemsPanelTemplate>
-      </Setter.Value>
-    </Setter>
-  </Style>
-</Styles>
\ No newline at end of file
+    <Style Selector="TabControl">
+        <Setter Property="Background" Value="{DynamicResource ThemeBackgroundBrush}"/>
+        <Setter Property="BorderBrush" Value="{DynamicResource ThemeBorderMidBrush}"/>
+        <Setter Property="BorderThickness" Value="{DynamicResource ThemeBorderThickness}"/>
```
Comment (grokys) [inline src/Avalonia.Themes.Default/TabControl.xaml]: Do we want to have a border around the `TabControl` by default? IMO it looks a little strange and as far as I know, UWP (which the theme is based on) doesn't have it.

**Review by grokys — CHANGES_REQUESTED** [review]
This is looking great, a massive improvement. Just a few small things I noticed.

File: src/Avalonia.Themes.Default/TabControl.xaml (hunk 5 above)
Comment (Gillibald) [reply src/Avalonia.Themes.Default/TabControl.xaml]: Removing it is fine for me will do so.

File: src/Avalonia.Controls/Generators/TabItemContainerGenerator.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Generators/TabItemContainerGenerator.cs]: Should be safe to remove that check 👍

**Comment by Gillibald** [comment]
@grokys Thanks for your review. Will make the requested changes in the next days. This helps me a lot to write better code for this project.

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Thanks @Gillibald! Merged!

**Pull request merged**